dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/event/EventOperationParamsMapper.java

```diff
@@ -236,22 +265,115 @@ private OrganisationUnit validateOrgUnit(String orgUnitUid) throws BadRequestExc
+    }
+  }
+
+  /**
+   * Returns a list of all the org units the user has access to
+   *
+   * @param program the program the user wants to access to
+   * @param user the user to check the access of
+   * @param orgUnit parent org unit to get descendants/children of
+   * @param orgUnitDescendants function to retrieve org units, in case ou mode is descendants
+   * @return a list containing the user accessible organisation units
+   */
+  private List<OrganisationUnit> getUserAccessibleOrgUnits(
+      Program program,
+      User user,
+      OrganisationUnit orgUnit,
+      OrganisationUnitSelectionMode orgUnitMode,
+      Function<String, List<OrganisationUnit>> orgUnitDescendants) {
+
+    if (orgUnitMode == null && orgUnit != null) {
+      return getSelectedOrgUnits(user, program, orgUnit);
+    }
+
+    if (orgUnitMode == null) {
+      return getAccessibleOrgUnits(user, program);
     }
 
-    if (ps != null && !user.isSuper() && !aclService.canDataRead(user, ps)) {
-      throw new ForbiddenException("User has no access to program stage: " + ps.getUid());
+    return switch (orgUnitMode) {
+      case DESCENDANTS -> orgUnit != null
+          ? getAccessibleDescendants(orgUnitDescendants.apply(orgUnit.getUid()), program, user)
+          : Collections.emptyList();
+      case CHILDREN -> orgUnit != null
+          ? getAccessibleDescendants(
+              Stream.concat(Stream.of(orgUnit), orgUnit.getChildren().stream()).toList(),
+              program,
+              user)
+          : Collections.emptyList();
+      case CAPTURE -> user.getOrganisationUnits().stream().toList();
+      case ACCESSIBLE -> getAccessibleOrgUnits(user, program);
+      case SELECTED -> getSelectedOrgUnits(user, program, orgUnit);
+      default -> Collections.emptyList();
+    };
+  }
+
+  private List<OrganisationUnit> getSelectedOrgUnits(
+      User user, Program program, OrganisationUnit orgUnit) {
+    return trackerAccessManager.canAccess(user, program, orgUnit)
+        ? List.of(orgUnit)
+        : Collections.emptyList();
+  }
+
+  private List<OrganisationUnit> getAccessibleOrgUnits(User user, Program program) {
+    return isProgramAccessRestricted(program)
+        ? user.getOrganisationUnits().stream().toList().stream().toList()
+        : user.getTeiSearchOrganisationUnitsWithFallback().stream().toList();
+  }
+
+  /**
+   * Returns the org units whose path is contained in the user search or capture scope org unit. If
+   * there's a match, it means the user org unit is at the same level or above the supplied org
+   * unit.
+   *
+   * @param availableOrgUnits the org units to check if the user has access to
+   * @param program the program the user wants to access to
+   * @param user the user to check the access of
+   * @return a list with the org units the user has access to
+   */
+  private List<OrganisationUnit> getAccessibleDescendants(
+      List<OrganisationUnit> availableOrgUnits, Program program, User user) {
```

Comment:
nit: `availableOrgUnits` does not add value IMHO

```suggestion
      User user, Program program, List<OrganisationUnit> orgUnits) {
```

I switched order to keep user first and program second as this method seems almost like an overloaded one of

`private List<OrganisationUnit> getAccessibleOrgUnits(User user, Program program) {`

so we stay consistent